Remove gotchas variants of `Font::DetermineLines()`.

LGM-Doyle opened the pull request:
This PR follows #1035

This removes the gotcha variant of Font::DetermineLines that repeatedly create and discard the computationally expensive TextElement vector.

adrianbroher reviewed (changes requested):
Looks good in general but the documentation is lacking.

LGM-Doyle committed:
Rename Font::FillTextElements to Font::ExpensiveParseTextToTextElements.

Also make the function public, to allow its use to be separated from
DetermineLines.

LGM-Doyle committed:
In ExpensiveParseFromTextToTextElements replace ignore_tags with format.

LGM-Doyle committed:
Make ExpensiveParseFromTextToTextElements() return a vector of TextElements.

LGM-Doyle committed:
Add comments to Font::ExpensiveParseFromTextToTextElements().

LGM-Doyle committed:
Remove hazardous variant of Font::TextExtent.

Font had a variant of TextExtent() that 'helpfully' internally called
ExpensiveParseFromTextToTextElements and DetermineLines, computing and
discarding text_elements each time they were called.  This variant ran
10 times slower than the variant with the pre-computed text elements.

When called during Render() of other tight loops this results in delays
long enough to be perceived by humans.

This commit removes this TextExtent() variant.

It results in more verbose code for hazardous operations of repeatedly
generating text elements, merely to find a text extent.  It will
prompt the programmer to move ExpensiveParseFromTextToTextElements out
of the loop.

LGM-Doyle committed:
Add default value for format in ExpensiveParseFromTextToTextElements.

LGM-Doyle committed:
Remove hazardous variant of Font::PreRenderText().

Font had a variant of PreRenderText() that 'helpfully' internally called
DetermineLines which in turn called
ExpensiveParseFromTextToTextElements, computing and discarding
text_elements each time they were called.

When called during Render() or other tight loops this results in
noticable slowdowns.

This commit removes this variant.

It results in more verbose code for hazardous operations.  It will
prompt the programmer to move ExpensiveParseFromTextToTextElements out
of tight loops.

LGM-Doyle committed:
Remove hazardous variant of Font::RenderText().

Font had a variant of RenderText() that 'helpfully' internally called
DetermineLines which in turn called
ExpensiveParseFromTextToTextElements, computing and discarding
text_elements each time they were called.

When called during Render() or other tight loops this results in
noticable slowdowns.

This commit removes this variant.

It results in more verbose code for hazardous operations.  It will
prompt the programmer to move ExpensiveParseFromTextToTextElements out
of tight loops.

LGM-Doyle committed:
Remove gotcha variants of Font::DetermineLines.

Font::DetermineLines had two variants that did not require pre-computed
text_elements.  These variants 'helpfully' computed and discarded text
elements each time they were called.  These variants ran 10 times slower
than the variant with the pre-computed text elements.

These variants were used in equally helpful variants of TextExtent,
RenderText and PreRenderText.

When called during Render() or other tight loops this results in
noticable slowdowns.

This commit removes these variants.

It results in more verbose code for hazardous operations.  It will
prompt the programmer to move ExpensiveParseFromTextToTextElements out
of tight loops.

LGM-Doyle committed:
Expand description of DetermineLines() for Review #1036.

LGM-Doyle committed:
In SaveFileDialog stop using 1 << 15 as widest possible text.

Thread ClientWidth() through row and free functions when computing text
widths.

LGM-Doyle committed:
In EncyclopediaDetailPanel stop using 1 << 15 as widest possible text.

Thread ClientWidth() through free functions when computing text widths.

LGM-Doyle committed:
Fix call function once to fetch format in BrowseInfoWnd.cpp for review #1036.

LGM-Doyle reviewed (commented):
Thanks for the review.

I don't think you needed to rebase the last few PRs since I have already rebased them to the previous PR in the series, unlike the previous 3 which were in parallel.  I have also rebased the next few PRs as a series.

LGM-Doyle committed:
Style change from review #1040.
